gcs/blockcf2/maxsize_test.go

```diff
@@ -0,0 +1,268 @@
+	// that may be added without limits to a transaction).
+	p2shSize := uint32(23)
+
+	tests := []struct {
+		name string
+		n    uint32
+		want uint64
+	}{{
+		// This test asserts the size of a filter when the source
+		// data for the filter fills an entire P2P message.
+		name: "filter from data that fills an entire P2P msg",
+		n:    wire.MaxMessagePayload / entrySizeForSize(wire.MaxMessagePayload),
+		want: 22541387,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// for the filter fills as many bytes as the maximum block
+		// payload size for a P2P message.
+		name: "filter from data that fills the max block size",
+		n:    wire.MaxBlockPayload / entrySizeForSize(wire.MaxBlockPayload),
+		want: 1174034,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// for the filter fills as many bytes as the maximum consensus
+		// enforced block size for mainnet.
+		name: "filter from data that fills the max mainnet block size",
+		n:    mainnetMaxBlockSize / entrySizeForSize(mainnetMaxBlockSize),
+		want: 352214,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// for the filter is a tx that has as many OP_RETURN outputs
+		// as necessary to fill a block.
+		name: "filter from tx filled with OP_RETURN outputs",
+		n:    wire.MaxBlockPayload / (11 + entrySizeForSize(wire.MaxBlockPayload)),
+		want: 251581,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// is a set of transactions that have 4 OP_RETURN outputs, as
+		// enforced by the standardness policy checks of the mempool.
+		name: "filter from standard OP_RETURN tx",
+		n:    wire.MaxBlockPayload / (minTxSize + (txOutSize+entrySizeForSize(wire.MaxBlockPayload))*4) * 4,
+		want: 108370,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// is a single transaction with as many P2SH outputs as
+		// necessary to fill the largest block of any network.
+		name: "filter from P2SH outputs tx",
+		n:    wire.MaxBlockPayload / (txOutSize + p2shSize),
+		want: 103593,
+	}, {
+		// This test asserts the size of a filter when the source data
+		// is a single transaction with as many P2SH outputs as
+		// necessary to fill the largest block on mainnet.
+		name: "filter from P2SH outputs tx on mainnet",
+		n:    mainnetMaxBlockSize / (txOutSize + p2shSize),
+		want: 31080,
+	}}
+
+	for i := range tests {
+		tc := tests[i]
+		t.Run(tc.name, func(t *testing.T) {
+			got := gcs.MaxFilterV2Size(B, M, tc.n)
+			if tc.want != got {
+				t.Logf("N: %d", tc.n)
+				t.Fatalf("Unexpected max filter size: got %d, want %d",
+					got, tc.want)
+			}
+		})
+	}
+}
+
+func fastReduce(x, N uint64) uint64 {
+	hi, _ := bits.Mul64(x, N)
+	return hi
+}
+
+// TestWorstMaxSize generates the largest possible cfilter for a given random
+// key.
+func TestWorstMaxSize(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping due to -test.short")
+	}
```

Comment:
This test is a bit on the slow side (about 15 seconds on my machine) which is why I gated it under `!testing.Short`. It's intention is to be more informational then actually assertive though: it basically shows how it's _really really_ hard to build a cfilter that maximizes a predetermined block size.

Reply:
Good call on gating it.

By the way, this is quite a nice test.  It's great to see the worst case actually exercised.  I'm sure it took you a bit to work it all out.